filesys/handler.go

```diff
@@ -0,0 +1,300 @@
+	Fd int `json:"fd"`
+}
+
+func (c *Close) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	err := syscall.Close(FdType(c.Fd))
+	if err != nil {
+		fa.doError(syscall.ENOSYS.Error(), "ENOSYS", w)
+		return
+	}
+	fa.okResponse(map[string]any{}, w)
+}
+
+type Rename struct {
+	From string `json:"from"`
+	To   string `json:"to"`
+}
+
+func (r *Rename) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	err := syscall.Rename(fixPath(r.From), fixPath(r.To))
+	if fa.handleError(w, err, true) {
+		return
+	}
+	fa.okResponse(map[string]any{}, w)
+}
+
+type Readdir struct {
+	Path string `json:"path"`
+}
+
+func (r *Readdir) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	entries, err := os.ReadDir(fixPath(r.Path))
+	if err != nil {
+		fa.doError(syscall.ENOSYS.Error(), "ENOSYS", w)
+		return
+	}
+	stringNames := make([]string, 0)
+	for _, entry := range entries {
+		stringNames = append(stringNames, entry.Name())
+	}
+	fa.okResponse(map[string]any{"entries": stringNames}, w)
+}
+
+type Lstat struct {
+	Path string `json:"path"`
+}
+
+type Read struct {
+	Fd       int  `json:"fd"`
+	Offset   int  `json:"offset"`
+	Length   int  `json:"length"`
+	Position *int `json:"position,omitempty"`
+}
+
+func (r *Read) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	if r.Offset != 0 {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+	if r.Position != nil {
+		_, err := syscall.Seek(FdType(r.Fd), int64(*r.Position), 0)
+		if err != nil {
+			fa.doError("not implemented", "ENOSYS", w)
+			return
+		}
+	}
+
+	buffer := make([]byte, r.Length)
+	read, err := syscall.Read(FdType(r.Fd), buffer)
+	if err != nil {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+	response := map[string]any{
+		"read":   read,
+		"buffer": base64.StdEncoding.EncodeToString(buffer[:read]),
+	}
+	fa.okResponse(response, w)
+
+}
+
+type Mkdir struct {
+	Path string `json:"path"`
+	Perm uint32 `json:"perm"`
+}
+
+func (m *Mkdir) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	err := syscall.Mkdir(fixPath(m.Path), m.Perm)
+	if err != nil {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+	fa.okResponse(map[string]any{}, w)
+}
+
+type Unlink struct {
+	Path string `json:"path"`
+}
+
+func (u *Unlink) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	err := syscall.Unlink(fixPath(u.Path))
+	if err != nil {
+		fa.doError("not implemented", "ENOSYS", w)
+		return
+	}
+	fa.okResponse(map[string]any{}, w)
+}
+
+type Rmdir struct {
+	Path string `json:"path"`
+}
+
+func (r *Rmdir) WriteResponse(fa *Handler, w http.ResponseWriter) {
+	err := syscall.Rmdir(fixPath(r.Path))
+	if fa.handleError(w, err, true) {
+		return
+	}
+	fa.okResponse(map[string]any{}, w)
+}
+
+func (fa *Handler) handleError(w http.ResponseWriter, err error, noEnt bool) bool {
+	if err == nil {
+		return false
+	}
+	if noEnt && os.IsNotExist(err) {
+		fa.doError(syscall.ENOENT.Error(), "ENOENT", w)
+	} else {
+		fa.doError(syscall.ENOSYS.Error(), "ENOSYS", w)
```

Comment:
Any reason to be ENOSYS? Also, we seem to be completely gobbling up the actual `err`. Could we pass that down or log that somewhere?

Reply:
ENOSYS was the error code previously used in the wasm_exec.js for all unimplemented file system calls. 

I think it still makes sense to use it as a generic "not implemented" error with the following rationale:

For each of the different platforms: linux, mac, and windows, the various `syscall.Xxxx` are handled differently, have different error codes, and for windows, are not POSIX at all. It would be quite a task to sort out what the appropriate error code would be based on the current running platform.  ENOENT ( effectively file not found  ) was the most important one to get right.

I'll see about adding logging for the `err` in doError and updating all callers to pass in the error for logging.

Reply:
Sounds good, thanks.

Reply:
Sounds good, thanks.